src/methods.jl

```diff
@@ -10,318 +19,438 @@ derivatives.
 const DEFAULT_CONDITION = 100
 
 """
-    add_tiny(x::Real)
+    FiniteDifferenceMethod{G<:AbstractVector, C<:AbstractVector}
 
-Add a tiny number, 10^{-20}, to `x`, preserving the type. If `x` is an `Integer`, it is
-promoted to a suitable floating point type.
-"""
-add_tiny(x::T) where {T<:Real} = x + convert(T, 1e-20)
-add_tiny(x::Integer) = add_tiny(float(x))
+A finite difference method.
 
-forward_grid(p::Int) = 0:(p - 1)
-backward_grid(p::Int) = (1 - p):0
-function central_grid(p::Int)
-    if isodd(p)
-        return div(1 - p, 2):div(p - 1, 2)
-    else
-        return vcat(div(-p, 2):-1, 1:div(p, 2))
-    end
+# Fields
+- `grid::G`: Multiples of the step size that the function will be evaluated at.
+- `q::Int`: Order of derivative to estimate.
+- `coefs::C`: Coefficients corresponding to the grid functions that the function evaluations
+    will be weighted by.
+- `bound_estimator::Function`: A function that takes in the function and the evaluation
+    point and returns a bound on the magnitude of the `length(grid)`th derivative.
+"""
+struct FiniteDifferenceMethod{G<:AbstractVector, C<:AbstractVector}
+    grid::G
+    q::Int
+    coefs::C
+    bound_estimator::Function
 end
```

Comment:
Lets make it possible to access the `bounds_estimator` field in a type stable way.

```suggestion
struct FiniteDifferenceMethod{G<:AbstractVector, C<:AbstractVector, E<:Function}
    grid::G
    q::Int
    coefs::C
    bound_estimator::E
end
```